### PR TITLE
[WIP] Cluster implementation

### PR DESCRIPTION
**Pull request by vkorn**
TBD: 
- [x] Host operations
- [ ] Logs from remote
- [x] Describe API
- [x] Call remote node upon calling `/leave` from master
- [ ] Stop/Start hass 
- [ ] Initialize cluster addon
- [x] Check that docker has `libssl-dev` and `python-dev` (required for `jwcrypto`)
- [x] Finalize if we have to leave `/supervisor/info[addons]` as is

Breaking changes: 
1. Any addons lists `installed` field now contains dictionary of `{node: version}`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -90,3 +90,10 @@ ENV/
 
 # pylint
 .pylint.d/
+
+# PyCharm/Idea
+.idea
+
+# Local test files
+run_master.sh
+run_slave.sh
```
Comment (pvizeli) [inline .gitignore]: Can you exlude the *.sh from the commit?

File: hassio/api/__init__.py [hunk 2]
```diff
@@ -126,6 +190,18 @@ def register_snapshots(self, snapshots):
...
+    def register_cluster(self, cluster):
+        """Register cluster function."""
+        api_cluster = APICluster(cluster, self.config, self.loop)
+
+        self.webapp.router.add_get('/cluster/info', api_cluster.info)
+        self.webapp.router.add_post('/cluster/switch_to_master',
+                                    api_cluster.switch_to_master)
+        self.webapp.router.add_post('/cluster/switch_to_slave',
+                                    api_cluster.switch_to_slave)
+        self.webapp.router.add_post('/cluster/{slug}/remove',
```
Comment (pvizeli) [inline hassio/api/__init__.py]: I think we shuld Name it `/cluster/switch/master`, `/cluster/switch/slave`

File: hassio/api/__init__.py [hunk 3]
```diff
@@ -136,24 +212,43 @@ def get_panel(request):
...
-            await self.server.wait_closed()
-        await self.webapp.shutdown()
 
-        if self._handler:
-            await self._handler.finish_connections(60)
-        await self.webapp.cleanup()
+class ClusterAPI(RestAPIBase):
+    """Cluster management REST API."""
+
+    def register_cluster_management(self, cluster):
```
Comment (pvizeli) [inline hassio/api/__init__.py]: We should remove the namespace `public`

File: hassio/api/__init__.py [hunk 4]
```diff
@@ -136,24 +212,43 @@ def get_panel(request):
...
+        api_cluster = APIClusterManagement(cluster, self.config, self.loop)
+
+        self.webapp.router.add_post('/cluster/public/register',
+                                    api_cluster.register_node)
+        self.webapp.router.add_post('/cluster/public/unregister',
+                                    api_cluster.unregister_node)
+        self.webapp.router.add_post('/cluster/public/ping',
+                                    api_cluster.ping)
+
+    def register_cluster_addons(self, cluster, addons, api_addons):
```
Comment (pvizeli) [inline hassio/api/__init__.py]: We should remove the namespace `public`

File: hassio/core.py [hunk 5]
```diff
@@ -166,6 +185,7 @@ def __init__(self, loop, config):
         # process stop tasks
         self.websession.close()
         await self.api.stop()
+        await self.cluster_api.stop()
```
Comment (pvizeli) [inline hassio/core.py]: use `await asyncio.wait([coro, coro], loop=self.loop)`

File: hassio/core.py [hunk 6]
```diff
@@ -142,8 +160,9 @@ def __init__(self, loop, config):
             # start addon mark as before
             await self.addons.auto_boot(STARTUP_BEFORE)
 
-            # run HomeAssistant
-            await self.homeassistant.run()
+            # run HomeAssistant if we're not on slave node
+            if self.cluster.is_master is True:
+                await self.homeassistant.run()
```
Comment (pvizeli) [inline hassio/core.py]: Don't stop hass at the moment, we do it with a second PR.

File: hassio/dock/homeassistant.py [hunk 7]
```diff
@@ -40,7 +40,7 @@ def _run(self):
 
         Need run inside executor.
         """
-        if self._is_running():
+        if self._is_running() or self.config.is_master is False:
```
Comment (pvizeli) [inline hassio/dock/homeassistant.py]: Don't stop hass at the moment, we do it with a second PR.

File: hassio/api/__init__.py (hunk 2 above)
Comment (pvizeli) [inline hassio/api/__init__.py]: I think it should be okay to have only `/cluster/master/{node}` we should also update  `/cluster/{slug}/remove` to `/cluser/{node}/remove`

File: hassio/cluster/__init__.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                ATTR_VERSION: HASSIO_VERSION,
+                ATTR_ARCH: self.config.arch,
+                ATTR_TIMEZONE: self.config.timezone,
+            }
+
+            with async_timeout.timeout(10, loop=self.websession.loop):
+                async with self.websession.post(url,
+                                                headers=headers,
+                                                json=data) as request:
+                    response = await request.json(content_type=None)
```
Comment (pvizeli) [inline hassio/cluster/__init__.py]: we have content type set for json, so you can remove this

File: hassio/cluster/__init__.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        if self.is_master is False:
+            return local
+
+        tasks = []
+        for node_ in self.known_nodes:
+            if node_.is_active:
+                tasks.append(node_.get_addons_list())
+        len_ = len(tasks)
+        if len_ > 0:
+            results, _ = await asyncio.shield(
```
Comment (pvizeli) [inline hassio/cluster/__init__.py]: we use shield only on `/api/*` objects

**Comment by pvizeli** [comment]
We should not break the exists API. I think we should it do in this way:
```
/supervisor/info
{
    ...
    "cluster": "bool"
    "addons": [
        {
            ...
            "node": "slug"
        },
    ],
}
```

File: hassio/api/__init__.py (hunk 3 above)
Comment (vkorn) [inline hassio/api/__init__.py]: Done

File: hassio/api/__init__.py [hunk 10]
```diff
@@ -126,6 +190,18 @@ def register_snapshots(self, snapshots):
...
+        """Register cluster function."""
+        api_cluster = APICluster(cluster, self.config, self.loop)
+
+        self.webapp.router.add_get('/cluster/info', api_cluster.info)
+        self.webapp.router.add_post('/cluster/leave',
+                                    api_cluster.switch_to_master)
+        self.webapp.router.add_post('/cluster/register',
+                                    api_cluster.switch_to_slave)
+        self.webapp.router.add_post('/cluster/{node}/leave',
+                                    api_cluster.remove_node)
```
Comment (pvizeli) [inline hassio/api/__init__.py]: We should name the object function like rest function name

File: hassio/api/addons.py [hunk 11]
```diff
@@ -48,24 +49,21 @@ def _extract_addon(self, request, check_installed=True):
 
         return addon
 
+    def _get_cluster_node(self, request):
+        """Return cluster node if it's known and active."""
+        node_slug = request.match_info.get('node')
+        if node_slug is None or node_slug == CLUSTER_NODE_MASTER:
+            return None
+        node = self.cluster.get_node(node_slug)
```
Comment (pvizeli) [inline hassio/api/addons.py]: Like schema they work on other module, we should call this only `get`

File: hassio/cluster/util.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    """Fixing node name."""
+    return node_name.replace(" ", "_")
+
+
+def get_public_cluster_url(ip_address, relative_url):
+    """Preparing URL for public cluster API."""
+    if relative_url[0] != "/":
+        relative_url = "/" + relative_url
+    return "http://{0}:{1}/" \
+           "cluster/public{2}".format(ip_address,
```
Comment (pvizeli) [inline hassio/cluster/util.py]: We have remove the public sufix?

**Comment by pvizeli** [comment]
https://github.com/home-assistant/hassio-build/commit/ac10143eb158e9ec433a2e5e9587d52c3e6d0337

File: hassio/cluster/util.py (hunk 12 above)
Comment (vkorn) [inline hassio/cluster/util.py]: Yep, forgot to remove old function, thx

File: API.md [hunk 13]
```diff
@@ -378,6 +380,50 @@ Output the raw docker log
...
+``` 
+
+- POST `/cluster/register`
+
+Registering this instance as a slave node in cluster
+ ```json
+{
+	"master_ip": "ip address of master node", 
+	"master_key": "master key", 
+	"node_name": "registration name"
```
Comment (pvizeli) [inline API.md]: use only `name` like other api call use and in context it is correct.

File: hassio/api/addons.py [hunk 14]
```diff
@@ -87,103 +85,141 @@ def reload(self, request):
...
+            return await asyncio.shield(addon.uninstall(), loop=self.loop)
+
+        return await node.addon_uninstall(request)
 
     @api_process
     async def start(self, request):
         """Start addon."""
-        addon = self._extract_addon(request)
+        node = self._get_cluster_node(request)
+        if node is None:
```
Comment (pvizeli) [inline hassio/api/addons.py]: change the style for ident with:
```
if node:
    return ...

...
```

File: hassio/config.py [hunk 15]
```diff
@@ -52,6 +53,7 @@
     vol.Optional(SECURITY_PASSWORD): vol.Coerce(str),
     vol.Optional(SECURITY_SESSIONS, default={}):
         {vol.Coerce(str): vol.Coerce(str)},
+    vol.Optional(CLUSTER_IS_MASTER, default=True): vol.Boolean(),
```
Comment (pvizeli) [inline hassio/config.py]: save that inside cluster.json

File: hassio/config.py [hunk 16]
```diff
@@ -277,3 +279,14 @@ def security_sessions(self, value):
...
+    @property
+    def is_master(self):
+        """Return flag indicating whether this node operates as master."""
+        return self._data.get(CLUSTER_IS_MASTER)
+
+    @is_master.setter
+    def is_master(self, value):
+        """Set operation mode."""
+        self._data[CLUSTER_IS_MASTER] = value
+        self.save()
```
Comment (pvizeli) [inline hassio/config.py]: move that also into cluster.json / ClusterManager

File: hassio/core.py [hunk 17]
```diff
@@ -130,8 +144,18 @@ def __init__(self, loop, config):
         )
 
         # start api
-        await self.api.start()
-        _LOGGER.info("Start hassio api on %s", self.config.api_endpoint)
+        _LOGGER.info("Starting API on %s. Hassio port %d. Cluster port %d",
+                     self.config.api_endpoint, HASSIO_API_PORT,
+                     HASSIO_PUBLIC_CLUSTER_PORT)
+        tasks, _ = await asyncio.wait([
```
Comment (pvizeli) [inline hassio/core.py]: Ignore the return value / we have allready a error inside log

File: hassio/cluster/util_rest.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        full_url = get_public_cluster_url(node_ip, url)
+        data = cluster_encrypt_json(json_obj, plain_key)
+        with async_timeout.timeout(10, loop=websession.loop):
+            async with websession.request(method, full_url, data=data,
+                                          headers=headers) as response:
+                if is_raw:
+                    return await cluster_decrypt_body(response, plain_key)
+                return await cluster_decrypt_body_json(response, plain_key)
+    # pylint: disable=broad-except
+    except Exception as err:
```
Comment (pvizeli) [inline hassio/cluster/util_rest.py]: should be `(asyncio.TimeoutError, aiohttp.ClientError)`

File: hassio/cluster/util.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+def cluster_public_api_process(method):
+    """Wrapping every possible error."""
+
+    async def wrap_api(api, *args, **kwargs):
+        """Return api information."""
+        try:
+            result = await method(api, *args, **kwargs)
+            return web.Response(body=result)
+        # pylint: disable=broad-except
+        except Exception as err:
```
Comment (pvizeli) [inline hassio/cluster/util.py]: We use `RuntimeError`

File: hassio/config.py (hunk 15 above)
Comment (vkorn) [inline hassio/config.py]: Right now hass dock depends on this data.

File: hassio/cluster/util.py (hunk 19 above)
Comment (vkorn) [inline hassio/cluster/util.py]: This wraps various possible errors, including jwcrypt ones, not from hassio only

**Comment by pvizeli** [comment]
Maby we should make a slug from node name (like other module, use last 8 byte of hash -> https://github.com/home-assistant/hassio/blob/dev/hassio/snapshots/util.py#L9) and inside hassio we use this slug for referenc. The UI can fetch the node list know the full detail if we referenc to the slug.

File: hassio/cluster/util.py (hunk 19 above)
Comment (pvizeli) [inline hassio/cluster/util.py]: okay, add this 👍

File: hassio/config.py (hunk 15 above)
Comment (pvizeli) [inline hassio/config.py]: Yeah, but we ignore hass for this PR now, I do it with next PR

**Pull request closed without merging**